File: Days 41 - 50/Day 44/GetTotalInversionsFromArray.cs
using System;
using System.Collections.Generic;

namespace DailyCodingProblem
{
	internal class Day44
	{
		private static int Main(string[] args)
		{
			List<int> numbers = new List<int>() { 2, 4, 1, 3, 5 };
			Console.WriteLine(GetTotalInversions(numbers));

			numbers = new List<int>() { 5, 4, 3, 2, 1 };
			Console.WriteLine(GetTotalInversions(numbers));

			Console.ReadLine();

			return 0;
		}

		private static int GetTotalInversions(List<int> numbers) => MergeSort(numbers).Item2;

		private static (List<int>, int) MergeSort(List<int> numbers)
		{
			if (numbers.Count <= 1)
			{
				return (numbers, 0);
			}

			int midPoint = numbers.Count / 2;

			return Merge(MergeSort(numbers.GetRange(0, midPoint)), MergeSort(numbers.GetRange(midPoint, numbers.Count - midPoint)));
		}

		private static (List<int>, int) Merge((List<int>, int) firstHalf, (List<int>, int) secondHalf)
		{
			int i = 0;
			int j = 0;
			List<int> mergedList = new List<int>(firstHalf.Item1.Count + secondHalf.Item1.Count);

			(List<int> firstList, int firstInversions) = firstHalf;
			(List<int> secondList, int secondInversions) = secondHalf;

			int totalInversions = firstInversions + secondInversions;

			while (i < firstList.Count && j < secondList.Count)
			{
				if (firstList[i] < secondList[j])
				{
					mergedList.Add(firstList[i]);
					i++;
				}
				else
				{
					mergedList.Add(secondList[j]);
					totalInversions += firstList.Count - i;
					j++;
				}
			}

			while (i < firstList.Count)
			{
				mergedList.Add(firstList[i]);
				i++;
			}

			while (j < secondList.Count)
			{
				mergedList.Add(secondList[j]);
				j++;
			}

			return (mergedList, totalInversions);
		}
	}
}

File: Days 041 - 050/Day 44/GetTotalInversionsFromArray.cpp
#include <iostream>
#include <utility>
#include <vector>

std::pair<std::vector<int>, unsigned int> Merge(const std::pair<std::vector<int>, unsigned int>& firstHalf, const std::pair<std::vector<int>, unsigned int>& secondHalf) noexcept
{
	unsigned int i = 0;
	unsigned int j = 0;

	std::vector<int> mergedVector;
	mergedVector.reserve(firstHalf.first.size() + secondHalf.first.size());

	auto[firstVector, firstInversions] = firstHalf;
	auto[secondVector, secondInversions] = secondHalf;

	unsigned int totalInversions = firstInversions + secondInversions;

	while ((i < firstVector.size()) && (j < secondVector.size()))
	{
		if (firstVector[i] < secondVector[j])
		{
			mergedVector.push_back(firstVector[i]);

			i++;
		}
		else
		{
			mergedVector.push_back(secondVector[j]);
			totalInversions += firstVector.size() - i;

			j++;
		}
	}

	while (i < firstVector.size())
	{
		mergedVector.push_back(firstVector[i]);

		i++;
	}

	while (j < secondVector.size())
	{
		mergedVector.push_back(secondVector[j]);

		j++;
	}

	return std::make_pair(mergedVector, totalInversions);
}

std::pair<std::vector<int>, unsigned int> MergeSort(const std::vector<int>& numbers) noexcept
{
	if (numbers.empty() || numbers.size() == 1)
	{
		return std::make_pair(numbers, 0);
	}

	unsigned int midpoint = numbers.size() / 2;

	return Merge(MergeSort(std::vector<int>(numbers.begin(), numbers.begin() + midpoint)), MergeSort(std::vector<int>(numbers.begin() + midpoint, numbers.end())));
}

unsigned int GetTotalInversions(const std::vector<int>& numbers) noexcept
{
	return MergeSort(numbers).second;
}

int main(int argc, const char* argv[])
{
	std::cout << GetTotalInversions({ 2, 4, 1, 3, 5 }) << "\n";
	std::cout << GetTotalInversions({ 5, 4, 3, 2, 1 }) << "\n";

	std::cin.get();

	return 0;
}